### PR TITLE
Collaborative editing: Make syncing a side-concern instead of a replacement for local state

### PR DESCRIPTION
**Pull request by chriszarate**
## What?

Instead of effectively replacing the core-data store with a "database" Yjs provider (IndexedDB), this change makes syncing a side-effect of the `core-data` store, without changing how the entity records are represented and treated within it.

> [!NOTE]  
> This is the first of a series of modular PRs related to collaborative editing derived from the work in https://github.com/WordPress/gutenberg/pull/72040. Not all PRs result in a fully functional editing experience.

## Why?

- The primary motivation for this change is that **not all entity record properties should be synced**. When all properties are not synced, the simplest path is to continue to represent the entity in the local store.
   - Properties like `guid`, `_links`, and `modified` should not be synced because we want the local codebase to remain authoritative over their values.
   - Properties like `content` should not be synced because we prefer to sync a derivative property (`blocks`) instead.
- A secondary motivation is architectural simplicity. The `core-data` store continues to function as it does now, and the sync package interacts with it as a (privileged) client.
- A database Yjs provider (IndexedDB) _can_ be a vehicle for offline editing, but this is not easily implemented when a separate database (WordPress's relational database) is considered the source of truth.

## How?

- In `editEntityRecord` (`core-data` actions), do not skip dispatching edits and adding undo records for entities with syncing enabled.
  - Kevin's PR [took the same approach](https://github.com/WordPress/gutenberg/pull/68483/files#diff 
- In `getEntityRecord` (`core-data` resolvers), do not conditionally load the entity record into the store, and invert the order so that the entity is loaded before "bootstrapping."
  - This code change is minor (unwrapping an if/else), but produces a large diff due to the changed indentation.
963b9796c15a1a44431b3386775e7fe05e5bfd1f96293a8e20f5debae3478bc8). 
- Make sure to call the "register" method on the sync provider.
- In the entity sync config, introduce a set of synced properties. This set is currently limited to `blocks` to help navigate syncing errors, but it will be expanded in future PRs.

## Testing Instructions

1. Check out this branch.
2. Enable the real-time collaboration experiment: Gutenberg > Experiments.
3. Open a post for editing in two different browser environments.
4. Use your browser's development tools to remove the post editor lock (see screencast in the next section).

Note: This provider uses requests to `admin-ajax.php` for signaling and the initial connection is delayed.

### Testing Instructions for Keyboard

No UI changes in this pull request.

## Screenshots or screencast 

https://github.com/user-attachments/assets/40d3c34b-ca2a-4e82-aee6-6e644d67e1b3

### CODE REVIEW

File: packages/core-data/src/entities.js [hunk 1]
```diff
@@ -313,6 +313,7 @@ async function loadPostTypeEntities() {
 			name
 		);
 		const namespace = postType?.rest_namespace ?? 'wp/v2';
+		const syncedProperties = new Set( [ 'blocks' ] );
```
Comment (youknowriad) [inline packages/core-data/src/entities.js:316]: I don't understand why we need this change here? It feels very ad-hoc?

Comment (youknowriad) [reply packages/core-data/src/entities.js:316]: Also what about post types that don't support "content" (so have no "blocks" property)

Comment (chriszarate) [reply packages/core-data/src/entities.js:316]: > I don't understand why we need this change here? It feels very ad-hoc?

It is very ad hoc:

> In the entity sync config, introduce a set of synced properties. This set is currently limited to `blocks` to help navigate syncing errors, but it will be expanded in future PRs.

"Syncing errors" is shorthand for all sorts of edge cases, but the most serious issues are:

- Rich text attributes need to target the `raw` property.
- The `status` field cannot be synced until it "matures" past `"auto-draft"`, otherwise timing issues can result in the post getting stuck.

I didn't address these in this PR because we have more robust solutions available in other PRs, which I will open shortly. But I can create targeted fixes in this PR if that is desired.

Comment (youknowriad) [reply packages/core-data/src/entities.js:316]: I don't feel like this addresses my question though. I guess what I'm asking is that before making "blocks" a synced property for a given post type entity, we should first ensure the post type in question supports "content" (which we can do in the supports array no?)

Comment (youknowriad) [reply packages/core-data/src/entities.js:316]: Anyway, it's not a huge concern though but we should be thinking about post types that don't have "content" or "blocks".

**Review by youknowriad — APPROVED** [review]

File: packages/core-data/src/entities.js (hunk 1 above)
Comment (chriszarate) [reply packages/core-data/src/entities.js:316]: Yes, also addressed in a future PR. You can see how we approached this in #72040 in this commit:

https://github.com/WordPress/gutenberg/blob/48c1ebe642c57c17cb8bcf0334da0725d3d2af09/packages/core-data/src/utils/crdt.ts#L476

Apologies, I will have them all open today and can give you better pointers.

**Comment by youknowriad** [comment]
Thanks for splitting the work into small PRs, this one is good to go :)

**Comment by youknowriad** [comment]
Note that I sent you an invite to join the Gutenberg team, which means you should be able to use branches instead of a fork and also can merge approved PRs

**Comment by youknowriad** [comment]
I invite you to check the different PR practices (labels, merging props...)
